### ISSUE TITLE
ambertools integration?

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cmake-buildscripts#1]
hey, do you want to test your changes with amber source code? 

check here: https://github.com/Amber-MD/ambertools-ci/blob/nightly/devtools/ci/load_functions.sh

**Comment by multiplemonomials** [comment]
I'll set this up soon.  However, I can't merge the cmake stuff into the amber master branch until Dan Roe accepts my pull request to cpptraj

**Comment by drroe** [comment]
Oh sorry - I thought the PR was still a WIP. I'll merge.

**Comment by multiplemonomials** [comment]
wait, I'm not done yet!  I still have to add the parser for cpptrajfiles and ambpdbfiles.  I'll try to do that today or tomorrow, and then we can get things merged.

**Comment by hainm** [comment]
oh boy. You change your mind so quickly.

**Comment by multiplemonomials** [comment]
Sorry, I was unclear in my earlier post.  What I meant was that I had to finish my work on that PR so that Dr. Roe could merge it.

**Comment by multiplemonomials** [comment]
I've now got the CI build working with both CircleCI and Travis, so I think I can close this.

**Comment by hainm** [comment]
@multiplemonomials what do you think about setup CI for this repo? 
`cmake-buildscripts`.

Seems to me that master branch is stable and we should only merge the PR if passing the build (and may be simple test).

**Comment by multiplemonomials** [comment]
Yeah, but it really sucks to have to run `mkrelease_at`, wait 10 minutes for bzip2 and another 10 for the upload, and then commit to `ambertools_cmake_ci`  to trigger the CI build every time I make a change.  I really, really wish we could get a CI setup that could clone the Amber repo directly.  

Or are you saying that you want to add this to your nightly build?   That would be great, though it would take a fair amount of modification to your build scripts.

**Comment by hainm** [comment]
> Yeah, but it really sucks to have to run mkrelease_at...

Given that you will only change $AMBERHOME/cmake folder (which is this repo), you don't need to do anything else besides
`wget http://ambermd.org/downloads/ambertools-dev/AmberTools18-dev.tar.gz` (updated nightly);
then copy this repo to replace the $AMBERHOME/cmake folder. Then do the normal build. Aim here is to make sure that the master is always stable (that is syncing with amber master). We would do all experimental stuff in 'feature' branch, then make a PR to 'master' to wait for green build.

> I really, really wish we could get a CI setup that could clone the Amber repo directly.

ah, it will be much more slower. The repo is about > 7 GB :D.

> Or are you saying that you want to add this to your nightly build? That would be great, though it would take a fair amount of modification to your build scripts.

I will take care of this. ambermd server is currently pretty slow if there are >= 5 concurrent downloads so I put this on hold for now.

PS: bzip2 is very slow so I make a Python script to update mkrelease_at to use gzip.

**Comment by multiplemonomials** [comment]
Actually, I usually make changes to the CMake scripts everywhere in Amber, not just this repo.  It actually makes things a lot worse if the `cmake-buildscripts` submodule and the scripts in the Amber source tree are mismatched.

>ah, it will be much more slower. The repo is about > 7 GB :D.
Can't we just do a bare clone?  That's a lot smaller, right?  Or *is* that the bare clone?

As you can see, I had to make a **ton** of changes to the CI scripts for CMake support.  The hardest part was getting CMake 3.1+ installed on each of the CI setups.

**Comment by hainm** [comment]
> Can't we just do a bare clone? That's a lot smaller, right? Or is that the bare clone?

it's bare clone from scratch. I always have impression that it would take forever.

**Comment by hainm** [comment]
> The hardest part was getting CMake 3.1+ installed on each of the CI setups.

If it's very hard for you, how user can manage that? :D

**Comment by hainm** [comment]
but it's only a suggestion, take or leave it.

**Comment by multiplemonomials** [comment]
> If it's very hard for you, how user can manage that? :D

The problem is that they're all running ancient Linux distros that make things harder to work with.  Travis was the difficult one because I have to add a custom APT repository, and then it turned out that the `cmake` package conflicts with `emacs-common` for some crazy reason, so I had to monkey around with the package manager to get it to work.

**Comment by multiplemonomials** [comment]
Oh, by the way, should there be an MPI version of `gbnsr6` I noticed that the code contains MPI `#ifdef`s, but the makefiles don't build it in parallel mode.

**Comment by hainm** [comment]
> Travis was the difficult one because I have to add a custom APT repository, and then it turned out that the cmake package conflicts with emacs-common for some crazy reason, so I had to monkey around with the package manager to get it to work.

cmake is very popular and so is travis. I think we are not in right direction yet.

**Comment by hainm** [comment]
> Oh, by the way, should there be an MPI version of gbnsr6 I noticed that the code contains MPI #ifdefs, but the makefiles don't build it in parallel mode.

should talk to its maintainer. But I think's MPI version is not well supported/tested (yet), so they turn it off.

**Comment by hainm** [comment]
Ia there a file in this repo so I can see minimal requirement to have cmake work? (E.g: version)

Sometimes I am too lazy to surf wiki on my phone.

**Comment by multiplemonomials** [comment]
> should talk to its maintainer

Who is its maintainer?

> cmake is very popular and so is travis. I think we are not in right direction yet.

Travis makes all CMake users very annoyed with their Mesozoic-era CMake versions, it's not just me.  See [here](https://genbattle.bitbucket.io/blog/2016/01/17/c++-travis-ci/), [here](https://github.com/travis-ci/travis-ci/issues/7437), and [here](https://github.com/travis-ci/travis-ci/issues/2030).  I agree it's annoying, but a huge percentage of the Amber build system is built around the interface libraries feature that was added in CMake 3.1, so I can't really roll it back.

**Comment by hainm** [comment]
Regarding maintainer: I don't know but I usually do

git log some_folder

**Comment by hainm** [comment]
Corresponding author here: http://people.cs.vt.edu/onufriev/

**Comment by hainm** [comment]
You might be surprised that I am using centos 5  with gcc4.6 to build amber binary. Aim here is to make the binary works every where and user does not need much.

**Comment by swails** [comment]
> I really, really wish we could get a CI setup that could clone the Amber repo directly. - @multiplemonomials 

We can, through Jenkins: http://www.jasonswails.com:8888 -- set up a recipe there (it's hosted on my Linux desktop in my basement).

> ah, it will be much more slower. The repo is about > 7 GB :D. - @hainm 

The .git directory size is really the only thing that matters, and that's usually quite a bit smaller.  Of course this varies, but my copies are 3 and 4 GB.

Anyway, the way Jenkins works is that it does a full clone *once*, then it reuses the already-checked-out repo and optionally restores it to its pristine shape (git clean -fxd; git checkout .) before doing a pull.  In this case, working from the git repo directly should actually be substantially cheaper and faster than always downloading the full tarball.  You can optionally have Jenkins delete and re-clone the whole thing, but there's no point when you can just clean the existing copy.

> Oh, by the way, should there be an MPI version of gbnsr6? I noticed that the code contains MPI #ifdefs, but the makefiles don't build it in parallel mode. - @multiplemonomials 

That's from Alexey's lab (as @hainm pointed out).  It originally was just a code-copy of PBSA (which has an MPI version).  The MPI code may be left-over copy-pasta from those origins or some internal-only copy that's not ready for prime-time.

Since it's only useful for single-point calculations (no gradients last I knew) and already quite fast, there's probably not much importance placed on getting a parallel version working if no effort was made by the original developers to begin with.

> Can't we just do a bare clone? That's a lot smaller, right? Or is that the bare clone? - @multiplemonomials 

Bare is not what you want.  A bare repo has no working directories or folders -- it's literally *just* the contents of the .git folder.  That's how the repos are stored in remotes that only serve as centralized version control (e.g., on gitosis or GitHub).  But it's still the full history.  What a clone does is copy the full contents of a "bare" repository (or the .git folder if you're cloning a non-bare repo), then it actually "computes" each file from those contents!

I think what you meant to refer to is a "shallow" clone.  A shallow clone will omit large portions of the history and basically only pull in the objects needed to compose the current working files.

That's great if you don't care about history (and will never need to search through it, e.g. via ``git-bisect``).  In this case, if you were forced to use some CI framework that *didn't* cache previous clones, then a shallow clone would be just the ticket.  (And you can define how "deep" to clone as well, so shallowness is configurable).

Whew!  Nothing like digging up a stale convo! :smiley:

**Comment by hainm** [comment]
> The .git directory size is really the only thing that matters, and that's usually quite a bit smaller. Of course this varies, but my copies are 3 and 4 GB.

hic, my `.git` is about 4.1 G. Not sure why.

**Comment by hainm** [comment]
btw, can't wait for the gitlab you showed.

**Comment by multiplemonomials** [comment]
GitLab? We're getting GitLab?

**Comment by hainm** [comment]
I believe @swails is suggesting/making demo for that. Hopefully we (him) can make it real. :d

**Comment by multiplemonomials** [comment]
@swails Can we make an entirety-of-Amber target on that Jenkins server?  Do you mind setting me up an account so that I can work on that?

**Comment by swails** [comment]
> GitLab? We're getting GitLab?

www.jasonswails.com:1234 -- I'm still waiting to hear back from Dave about what he thinks -- feel free to play around with it but keep in mind that when I do convince someone to host it at git.ambermd.org, this instance will likely disappear.

> @swails Can we make an entirety-of-Amber target on that Jenkins server? Do you mind setting me up an account so that I can work on that?

I gave you an account on that Jenkins instance and created an `Amber` folder for you to play in.  You can create a job there to do whatever you want to try.

Your username for both resources is the same as your GH handle (multiplemonomials).  E-mail me for your password.

**Comment by multiplemonomials** [comment]
@swails Any idea why the Git clone is failing?  It seems like the key is wrong but it worked the first two times and no subsequent times.
```
Cloning the remote Git repository
Cloning repository gitosis@git.ambermd.org:amber
 > git init /var/lib/jenkins/home/workspace/Amber/AmberCI # timeout=10
Fetching upstream changes from gitosis@git.ambermd.org:amber
 > git --version # timeout=10
using GIT_SSH to set credentials Jamie's private key
 > git fetch --tags --progress gitosis@git.ambermd.org:amber +refs/heads/*:refs/remotes/origin/* # timeout=60
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --tags --progress gitosis@git.ambermd.org:amber +refs/heads/*:refs/remotes/origin/*" returned status code 128:
stdout: 
stderr: ssh_exchange_identification: read: Connection reset by peer
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1924)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1643)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:71)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:352)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:559)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1075)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1115)
	at hudson.scm.SCM.checkout(SCM.java:495)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1212)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:566)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:491)
	at hudson.model.Run.execute(Run.java:1729)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:419)
ERROR: Error cloning remote repo 'origin'
Finished: FAILURE
```

**Comment by swails** [comment]
I think the repeated failed clone attempts from Jenkins blacklisted my IP address -- cloning doesn't work from my house anymore.  I'm working on getting this resolved.

But it turns out my power company doesn't deliver as reliable a power supply as I'd like -- my machine lost power and is not coming back up.  I won't be able to investigate to fix until I get home tonight, so the Jenkins machine and GitLab POC will be down today.

**Comment by multiplemonomials** [comment]
That's odd because the clone only got interrupted (killed by the timeout) twice before it started failing in this manner.  Does the Amber git server really blacklist you after two failed attempts?

**Comment by swails** [comment]
> That's odd because the clone only got interrupted (killed by the timeout) twice before it started failing in this manner. Does the Amber git server really blacklist you after two failed attempts?

I was indeed blacklisted.  I've since been added to a permanent whitelist so this shouldn't happen again.

Also, I got home and restarted git, Jenkins, and the executor slaves.

**Comment by swails** [comment]
Also it was a lot more than 2 requests -- Jenkins does a lot of git server querying in the background (for instance just setting up the job, Jenkins pinged git.ambermd.org several times to see if the server was reachable.

It's not hard to imagine a naive observer confusing Jenkins activity for a nefarious actor attacking the system.

**Comment by multiplemonomials** [comment]
Ah, I see.  Sorry about that.

**Comment by swails** [comment]
By the way, do *not* do a "wipe out repository and force clone".  That's it needlessly consumes a ton of bandwidth and takes much longer than it has to.

Just do a "clean after checkout" to get exactly the same starting state.

**Comment by multiplemonomials** [comment]
Yeah, I just did that as a one-time thing to see if it would help the clone error.  I meant to turn it off once it was fixed, and I just did.

**Comment by swails** [comment]
:) just wanted to make sure
